### PR TITLE
Improved Diff type to use for text diff

### LINKED ISSUE #57: Display value diffs as text diffs
Before diving deeper in the range of changes we can properly detect, it would be good to be able to display the changes we already detect.

We now have a basic representation of changes to values. We would like to be able to pretty print such changes in a text diff-like format for human consumption.

The goal here is to add functions to the library that can print our diff as a text diff and to update the binary diff tool so that it outputs such diffs instead of simply printing `API changed!`.

### PR DESCRIPTION
**Pull request by Siddhi-agg**
Fixes #57.

This PR aims to make the following changes:
- Improves the `diff` type to incorporate value descriptions of the added/removed values
- Improves the `diff` type to incorporate old and new value descriptions of the modified values
- Resolves the aliases in value descriptions added to the `diff` type using the current `Env.t`

There is still work left to be done for #57, which will allow us to print a text diff from the `diff` type returned by `diff_interface` function.

### CODE REVIEW

**Review by NathanReb — COMMENTED** [review]
It looks good! I'm having doubts though regarding our expansion here and whether it will work as we expect. It would be nice to make sure of it.

How about adding a test that checks that the value_description type is properly expanded, that would clear things.

I also don't think it's necessary to expand the types when a value is added or removed, as it is likely to confuse users more than helping them. Ideally, we would like to expand the type only if it's necessary for the diff to be readable but that's a bit hard to do with our current representation and probably not worth the effort as I expect having a more precise diff represenation will make that trivial.
I suggest that for now we stick to only expanding types in the `Modified` case. What do you think?

**Comment by Siddhi-agg** [comment]
> Ideally, we would like to expand the type only if it's necessary for the diff to be readable but that's a bit hard to do with our current representation and probably not worth the effort as I expect having a more precise diff represenation will make that trivial.

Yeah, working on a better representation for diffs seems the better way to go, rather than trying to determine if expanding the type is needed for a more readable diff. 

Also, I think we can go for resolving type expressions only for the `Modified` types. I thought resolving it also in case of `Added` and `Removed` types would be consistent and thus easy for the users to understand. But what you say makes sense, so we can go ahead with that.


> How about adding a test that checks that the value_description type is properly expanded, that would clear things.

Yeah, that's a good idea! I will work on that.

**Review by NathanReb — COMMENTED** [review]
The new test looks great!

I added some comments on the expansion function that needs to support more cases to be complete but other than that it looks good!

File: lib/api_watch_diff.ml [hunk 1]
```diff
@@ -1,14 +1,33 @@
...
+  match get_desc typ with
+  | Tconstr _ -> Ctype.expand_head env typ
+  | Tarrow (label, param_type, result_type, commutable) ->
+      let expanded_param = expand env param_type in
+      let expanded_result = expand env result_type in
+      if expanded_param == param_type && expanded_result == result_type then typ
+      else
+        Ctype.newty
+          (Tarrow (label, expanded_param, expanded_result, commutable))
+  | _ -> typ
```
[the remark this answers is not in the document]
Comment (Siddhi-agg) [reply lib/api_watch_diff.ml:29]: Sure, I will add more cases for that!

File: lib/api_watch_diff.ml [hunk 2]
```diff
@@ -1,14 +1,33 @@
...
 
 module FieldMap = Map.Make (struct
   type t = string
 
   let compare = String.compare
 end)
 
+let rec expand (env : Env.t) (typ : Types.type_expr) : Types.type_expr =
+  match get_desc typ with
+  | Tconstr _ -> Ctype.expand_head env typ
```
[the remark this answers is not in the document]
Comment (Siddhi-agg) [reply lib/api_watch_diff.ml:21]: @NathanReb I think `expand_head` function deals with that using `expand_abbrev`. Is that not the case?

[the remark this answers is not in the document]
Comment (NathanReb) [reply lib/api_watch_diff.ml:21]: I'd need to check again but I would expect it not to!

[the remark this answers is not in the document]
Comment (NathanReb) [reply lib/api_watch_diff.ml:21]: It's also pretty easy to test for, use the following interface and swap the type `t` is aliasing and inspect the diff result.
```
type t = int

val x : (t, string) result
```
You can also simply try to `expand_head` on the type expr for `x` directly and see what you get there.

File: lib/api_watch_diff.ml [hunk 3]
```diff
@@ -1,7 +1,14 @@
 open Types
 
-type change = Added | Removed | Modified
-type diff = Value of string * change | Any
+type ('item, 'diff) change =
+  | Added of 'item
+  | Removed of 'item
+  | Modified of 'diff
```
Comment (NathanReb) [inline lib/api_watch_diff.ml:6]: For now we can probably simplify it as:
```suggestion
type ('item, 'diff) change =
  | Added of 'item
  | Removed of 'item
  | Modified of {ref: 'item; current: 'item}
```

**Review by NathanReb — COMMENTED** [review]
Looks good! I just made a suggestion to simplify the diff type for now but then we're good to go!

**Pull request merged**